From 38c9f4806f66f9e847d44d1ea414254382825475 Mon Sep 17 00:00:00 2001
From: Jean-Christophe PINCE <jean-christophe.pince@intel.com>
Date: Fri, 12 Oct 2012 13:01:17 +0200
Subject: [PORT FROM MAIN] Prevents the player to notify the client after disconnecting

BZ: 59148

The method MediaPlayerService::Client::disconnect needs to set the
notification callback prior to clearing the mClient attribute.
This is to prevent the situation when the notification occurs during
the disconnection.

Change-Id: I3b0d824a3e62e2b5d944da364db4222527df414e
Orig-Change-Id: I40264181427b5939f71c3c1816ff872315e5e22d
Signed-off-by: Jean-Christophe PINCE <jean-christophe.pince@intel.com>
Reviewed-on: http://android.intel.com:8080/81059
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: M, Arulselvan <arulselvan.m@intel.com>
Tested-by: M, Arulselvan <arulselvan.m@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 media/libmediaplayerservice/MediaPlayerService.cpp |   10 ++++++----
 1 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/media/libmediaplayerservice/MediaPlayerService.cpp b/media/libmediaplayerservice/MediaPlayerService.cpp
index 7ce6d1f..6c1fe26 100644
--- a/media/libmediaplayerservice/MediaPlayerService.cpp
+++ b/media/libmediaplayerservice/MediaPlayerService.cpp
@@ -530,15 +530,17 @@ void MediaPlayerService::Client::disconnect()
         Mutex::Autolock l(mLock);
         p = mPlayer;
     }
+    // clear the notification to prevent callbacks to dead client
+    // and reset the player. We assume the player will serialize
+    // access to itself if necessary.
+    if (p != 0)
+        p->setNotifyCallback(0, 0);
+
     mClient.clear();
 
     mPlayer.clear();
 
-    // clear the notification to prevent callbacks to dead client
-    // and reset the player. We assume the player will serialize
-    // access to itself if necessary.
     if (p != 0) {
-        p->setNotifyCallback(0, 0);
 #if CALLBACK_ANTAGONIZER
         ALOGD("kill Antagonizer");
         mAntagonizer->kill();
-- 
1.7.4.1

